### PR TITLE
feat: customize Electron cache

### PR DESCRIPTION
**Pull request by emmkimme**
#### Description of Change
This MR covers the following issues:
[Customize Electron/Chromium cache location](https://github.com/electron/electron/issues/18309)
[AppData folder has large cache which affects "roaming profile" users](https://github.com/desktop/desktop/issues/6523)

Notes: Allow to customize Electron/Chromium cache location

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are unchanged 
- [x] relevant documentation is changed or added
- [x] PR title follows semantic
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Technical Notes
The notion of cache (DIR_CACHE, DIR_USER_CACHE) is already present in the code but not used.

atom\browser\api\atom_api_app.cc
```
int GetPathConstant(const std::string& name) {
...
  else if (name == "cache")
    return DIR_CACHE;
  else if (name == "userCache")
    return DIR_USER_CACHE;
...
```

atom/browser/atom_paths.h
```
...
  DIR_USER_CACHE,              // Directory where user cache can be written.
#if defined(OS_POSIX)
  DIR_CACHE = base::DIR_CACHE,  // Directory where to put cache data.
#else
  DIR_CACHE = base::DIR_APP_DATA,
#endif
...
```
lib\browser\api\app.ts
```
app._setDefaultAppPaths = (packagePath) => {
...
  app.setPath('userData', path.join(app.getPath('appData'), app.name!))
  app.setPath('userCache', path.join(app.getPath('cache'), app.name!))
```

There is 2 levels of cache : 
- userCache (DIR_USER_CACHE) and cache (DIR_CACHE) 
equivalent to the 2 levels of user data
- userData (DIR_USER_DATA) and appData (DIR_APP_DATA) 

Purpose was to use the userCache value for customizing the cache

atom\browser\atom_browser_context.cc
Before
```
if (!base::PathService::Get(DIR_USER_DATA, &path_)) {
    base::PathService::Get(DIR_APP_DATA, &path_);
    path_ = path_.Append(base::FilePath::FromUTF8Unsafe(GetApplicationName()));
    base::PathService::Override(DIR_USER_DATA, path_);
  }
```

After
```
  if (!base::PathService::Get(DIR_USER_CACHE, &path_)) {
    base::PathService::Get(DIR_CACHE, &path_);
    path_ = path_.Append(base::FilePath::FromUTF8Unsafe(GetApplicationName()));
    base::PathService::Override(DIR_USER_CACHE, path_);
  }
```

#### Usage
```
import { app } from 'electron';

// Must be called before application is ready
app.setPath('userCache', myOwnUserCache);
```

#### Additional notes
For concurrent access violation, the cache folder must not be shared by different users.
For performance purpose, it must be set on an efficient drive.

We can even imagine to got it deep further and have a DIR_CACHE independent as well. So user can either change the cache (and let the system append the application name) or change the userCache

### CODE REVIEW

**Comment by emmkimme** [comment]
I restore files ending (I'm always in pain with that).
You are right, it will not work anymore for people using userData for moving the cache., I did not realized. 
But using userData for changing the Electron Cache location was a desperate solution.
To not break this, I can define userCache empty by default and fallback to previous implementation in this case.
Do you prefer ?

File: atom/browser/atom_browser_context.cc [hunk 1]
```diff
@@ -78,10 +78,10 @@ AtomBrowserContext::AtomBrowserContext(const std::string& partition,
   base::StringToInt(command_line->GetSwitchValueASCII(switches::kDiskCacheSize),
                     &max_cache_size_);
 
-  if (!base::PathService::Get(DIR_USER_DATA, &path_)) {
-    base::PathService::Get(DIR_APP_DATA, &path_);
+  if (!base::PathService::Get(DIR_USER_CACHE, &path_)) {
+    base::PathService::Get(DIR_CACHE, &path_);
     path_ = path_.Append(base::FilePath::FromUTF8Unsafe(GetApplicationName()));
-    base::PathService::Override(DIR_USER_DATA, path_);
+    base::PathService::Override(DIR_USER_CACHE, path_);
```
Comment (ckerr) [inline atom/browser/atom_browser_context.cc]: The docs indicate that the intent is for `DIR_USER_DATA` and `DIR_USER_CACHE` to _both_ exist side-by-side, but this PR removes the code that sets the fallback value of `DIR_USER_DATA`. Is that intentional?

File: docs/api/app.md [hunk 2]
```diff
@@ -641,9 +643,9 @@ created by this method. On failure an `Error` is thrown.
 
 You can only override paths of a `name` defined in `app.getPath`.
 
-By default, web pages' cookies and caches will be stored under the `userData`
+By default, web pages' cookies and caches will be stored under the `userCache`
```
Comment (ckerr) [inline docs/api/app.md]: Is this actually true? Does overriding `DIR_USER_CACHE` in the AtomBrowserContext constructor actually change behavior s.t. all the cache data goes there?

File: atom/browser/atom_browser_context.cc (hunk 1 above)
Comment (ckerr) [inline atom/browser/atom_browser_context.cc]: If we're setting `DIR_USER_CACHE` we probably also want to ensure it exists, as we do for `DIR_USER_DATA` in `Browser::DidFinishLaunching()`?

File: docs/api/app.md [hunk 3]
```diff
@@ -598,6 +598,8 @@ You can request the following paths by the name:
   * `~/Library/Application Support` on macOS
 * `userData` The directory for storing your app's configuration files, which by
   default it is the `appData` directory appended with your app's name.
+* `userCache` The directory for storing Electron cache configuration files, which by
```
Comment (ckerr) [inline docs/api/app.md]: s/cache configuration/cache

File: atom/browser/atom_browser_context.cc (hunk 1 above)
Comment (emmkimme) [reply atom/browser/atom_browser_context.cc]: At startup DIR-CACHE is set by default to DIR_USER_DATA, so the fallback may remain. The problem is when the userData is changed, the DIR_CACHE is not sync and the cache does not follow anymore, It is what @MarshallOfSound want to fix. I agree and try to find the best solution

File: docs/api/app.md (hunk 2 above)
Comment (emmkimme) [reply docs/api/app.md]: This notion of DIR_CACHE, DIR_USER_CACHE already exist in the code but are R/O info, changing them have currently no effect.

File: docs/api/app.md (hunk 3 above)
Comment (emmkimme) [reply docs/api/app.md]: thanks will update

**Comment by emmkimme** [comment]
I have a solution which should keep the backward compatibility, But unfortunately I have to introduce an additional path 'appCache'.

File: docs/api/app.md [hunk 4]
```diff
@@ -604,6 +604,8 @@ You can request the following paths by the name:
   * `~/Library/Application Support` on macOS
 * `userData` The directory for storing your app's configuration files, which by
   default it is the `appData` directory appended with your app's name.
+* `appCache` The directory for storing your app's Electron cache, which by
+  default it is the `appData` directory appended with your app's name.
```
Comment (zcbenz) [inline docs/api/app.md]: Document should warn user that setting `appCache` would make app incompatible with old data and they need to migrate data from old dir to new dir themselves.

Comment (zcbenz) [reply docs/api/app.md]: Using `appCache` as key is quite confusing as `app`-prefix has other meanings in `appData`. Maybe `browserCache` is better.

**Comment by zcbenz** [comment]
I'm good with introducing an opt-in feature to change cache directory, considering the complexity of correctly migrating old data, existing apps might not be willing to use it, but I think it is fine to give new apps an option.

/cc @MarshallOfSound @ckerr as the PR has updated.

**Comment by miniak** [comment]
@emmkimme please rebase

**Comment by codebytere** [comment]
@emmkimme please address comments and rebase, or this PR will be closed in 5 days due to inactivity.

**Comment by emmkimme** [comment]
Hi, I have a better solution in mind which should handle backward. I will delete the branch and redo from scratch. Thks for your comments.

**Pull request closed without merging**